lingpy/sequence/ngrams.py

```diff
@@ -0,0 +1,1441 @@
+# *-* coding: utf-8 *-*
+"""
+This modules provides methods for generating and collecting ngrams.
+
+The methods allow to collect different kind of subsequences, such as standard
+ngrams (preceding context), skip ngrams with both single or multiple gap
+openings (both preceding and following context), and positional ngrams (both
+preceding and following context).
+"""
+
+from __future__ import division
+import math
+import operator
+import pickle
+import random
+from collections import defaultdict, Counter
+from itertools import chain, combinations, product
+from functools import partial
+from lingpy.sequence.smoothing import smooth_dist
+
+# Global padding symbol, shared across all functions/class-methods.
+_PAD_SYMBOL = '$$$'
+_ELM_SYMBOL = '###'
+
+def _seq_as_tuple(sequence):
+    """
+    Internal function for automatically converting a string sequence to a
+    tuple, if needed.
+
+    Parameters
+    ----------
+    sequence: list or str
+        The sequence that shall be converted into an iterable.
+
+    Returns
+    -------
+    out: tuple
+        A tuple of the sequence.
+    """
+
+    # We first check for datatype and then for a space, as the first test is
+    # faster (and evaluation is lazy).
+    if isinstance(sequence, str) and ' ' in sequence:
```

Comment:
If we still aim at py2 compat (or should we drop it?), this would need to be `six.string_types`, I guess.

Reply:
Yes, I'd say: let's stick to py2 compat for the time being. It's no longer difficult to do.

Reply:
Oh yes, I am trying to keep it py2 compatible, only loading `six` is a no-brainer.